### PR TITLE
[r11s] Leveraging NetworkError to improve HTTP communication error handling

### PR DESCRIPTION
**Pull request by hedasilv**
Currently, r11s marks documents as corrupted in situations where no fatal error occurred. We aim to improve that, and the first step is to allow HTTP error handling to have more helpful information available for different components that use HTTP communication. One problem is that our `BasicRestWrapper` only adds the HTTP error status code to the response upon detecting an HTTP error. This PR updates `BasicRestWrapper` to instead rely on `NetworkError`, which represents errors that can happen due to network calls/HTTP. That is similar to the approach adopted by the r11s driver.

`NetworkError` is helpful in that it provides more information than just the status code, such as whether the error is fatal and retry-able, whether there is a message and a delay before the request can be retried. `BasicRestWrapper` leverages `NetworkError` to provide components with those important pieces of information. The `BasicRestWrapper`, which is built on top of `axios`, was also updated to take into account network failure scenarios, where the call fails due to networking issues.

### CODE REVIEW

File: server/routerlicious/api-report/server-services-client.api.md [hunk 1]
```diff
@@ -452,9 +467,17 @@ export interface IWriteSummaryResponse {
...
+    code: number, message: string, canRetry?: boolean, isFatal?: boolean, retryAfterMs?: number);
+    // (undocumented)
+    readonly canRetry?: boolean;
     readonly code: number;
+    // (undocumented)
+    get details(): INetworkErrorDetails | string;
+    // (undocumented)
+    readonly isFatal?: boolean;
+    // (undocumented)
+    readonly retryAfterMs?: number;
```
Comment (tylerbutler) [inline server/routerlicious/api-report/server-services-client.api.md:480]: Can you document these new properties and methods as you add them?

Comment (hedasilv) [reply server/routerlicious/api-report/server-services-client.api.md:480]: @tylerbutler should I add the comments on server-services-client.api.md or directly in the source file that generated the report?

Comment (znewton) [reply server/routerlicious/api-report/server-services-client.api.md:480]: Source file, the api.md is auto-generated I believe

Comment (hedasilv) [reply server/routerlicious/api-report/server-services-client.api.md:480]: Got it. I was just reading about the [API Extractor](https://api-extractor.com/pages/tsdoc/doc_comment_syntax/). Will add the comments there, make sure the API docs get extracted and push a new iteration.

Comment (hedasilv) [reply server/routerlicious/api-report/server-services-client.api.md]: I added the comments for TS Doc, and I think the API report is showing that all new properties/methods/classes are now documented. Let me know if there is anything else to fix here regarding documentation. Thanks!

**Review by tanviraumi — APPROVED** [review]
Looks good to me. @znewton should take a look as well

File: server/routerlicious/packages/services-client/src/error.ts [hunk 2]
```diff
@@ -3,15 +3,169 @@
...
+ */
+export interface INetworkErrorDetails {
+    canRetry: boolean;
+    isFatal: boolean;
+    message?: string;
+    retryAfter?: number;
+}
+
+/**
+ * Represents errors associated with network communication communication.
```
Comment (tylerbutler) [inline server/routerlicious/packages/services-client/src/error.ts:17]:
```suggestion
 * Represents errors associated with network communication.
```

**Review by tylerbutler — APPROVED** [review]
@hedasilv Thanks a lot for adding docstrings. It dramatically improves the downstream dev experience, and is easier to do when writing the code than later. :)

**Review by sharmakhushboo — CHANGES_REQUESTED** [review]

File: server/routerlicious/api-report/server-services-client.api.md [hunk 3]
```diff
@@ -305,6 +308,18 @@ export interface IHistorian extends IGitService {
     getHeader(sha: string): Promise<resources.IHeader>;
 }
 
+// @public
+export interface INetworkErrorDetails {
+    // (undocumented)
+    canRetry: boolean;
+    // (undocumented)
+    isFatal: boolean;
```
[the remark this answers is not in the document]
Comment (hedasilv) [reply server/routerlicious/api-report/server-services-client.api.md]: Discussed offline. It's true that fatal errors cannot be retried. But a non-fatal error can either be retry-able or not. That is why we keep both `canRetry` and `isFatal`. In other words, we have 3 different scenarios:

a. errors that cause non-recoverable impact. (by default, cannot be retried)
b. errors that do not cause non-recoverable impact and can be retried;
c. errors that do not cause non-recoverable impact but cannot be retried;

To tell a. and c. apart, we use `isFatal`. To make it easier to understand what those different properties mean, I added more descriptive comments to the source files.

**Review by sharmakhushboo — APPROVED** [review]

File: server/routerlicious/packages/services-client/src/error.ts [hunk 4]
```diff
@@ -3,15 +3,186 @@
...
+
+/**
+ * Convenience function for generating a {@link NetworkError}.
+ * @remarks Generates a {@link NetworkError} instance appropriately configured given the status code and error data
+ * provided. This function is intended to be used in situations where a {@link NetworkError} is dynamically created
+ * based variable parameters. That is, when it is not known whether the status code can be 404 or 500.
+ * @param statusCode - HTTP status code that describes the error.
+ * @param errorData - Optional additional data associated with the error. Can either be a simple string representing
+ *  the message, or an {@link INetworkErrorDetails} object.
+ * @returns A simple string conveying the message if no other details are included in this {@link NetworkError},
```
Comment (znewton) [inline server/routerlicious/packages/services-client/src/error.ts:102]: nit: it looks like it always throws a NetworkError, never a string

File: server/routerlicious/packages/services-client/src/restWrapper.ts [hunk 5]
```diff
@@ -147,10 +153,26 @@ export class BasicRestWrapper extends RestWrapper {
...
                     } else {
-                        reject(error);
+                        if (error?.response?.status) {
+                            reject(createFluidServiceNetworkError(error?.response?.status, error?.response?.data));
+                        } else if (error?.message === "Network Error") {
+                            // If a service is temporarily down or a browser resource limit is reached, Axios will throw
+                            // a network error with no status code (e.g. err:ERR_CONN_REFUSED or err:ERR_FAILED) and
+                            // error message, "Network Error". "Network Error" can be retried, and is not fatal.
+                            reject(new NetworkError(
+                                -1, /* code */
```
Comment (znewton) [inline server/routerlicious/packages/services-client/src/restWrapper.ts:164]: If this bubbles up to another BasicRestWrapper, lets say Scribe calls Historian calls GitRest, and GitRest connection is down, Historian would send an `HTTP -1 "Network Error"` to Scribe's BasicRestWrapper.
1. Does Express.js allow setting status to -1?
2. When Scribe's BasicRestWrapper gets to line 158 and goes into `createFluidServiceNetworkError`, the switch case will result in `isFatal=true` and `canRetry=false`, but we would want to retry this type of error, right? I believe a quick fix would be to allow `isFatal` and `canRetry` variables to override their params in the default case of the switch/case

File: server/routerlicious/packages/services-client/src/error.ts (hunk 4 above)
Comment (hedasilv) [reply server/routerlicious/packages/services-client/src/error.ts:102]: Nice catch! It was a copy/paste failure from my side :) let me fix that.

File: server/routerlicious/packages/services-client/src/restWrapper.ts (hunk 5 above)
Comment (hedasilv) [reply server/routerlicious/packages/services-client/src/restWrapper.ts]: Thanks for noticing that. I updated the error handling logic in BasicRestWrapper (based on this https://axios-http.com/docs/handling_errors). Now I don't use -1 anymore. Also, the new error handling setup will correctly take care of the network errors, which as we discussed offline, could surface through a different message (ENOTFOUND)

**Review by znewton — APPROVED** [review]

**Pull request merged**